File: ui/src/types/backtester.ts
export interface ToolParam {
  name: string
  value: number
  min: number
  max: number
  step: number
}

export interface BacktestTool {
  idx: number
  name: string
  tooltip: string
  params: ToolParam[]
}

export interface StrategyEntry {
  id: string
  name: string
  domain: string
}

export interface DataSource {
  id: string
  name: string
  domain: string
}

export interface BacktestTradeRow {
  strategy_id: string
  side: string
  entry_price: number
  exit_price: number
  size: number
  pnl: number
  timestamp: number
}

export interface BacktesterState {
  strategies: StrategyEntry[]
  data_sources: DataSource[]
  tools: BacktestTool[]
  is_running: boolean
}

export interface BacktesterResults {
  equity_curve: number[]
  drawdown_curve: number[]
  pnl_buckets: [number, number][]
  mc_paths: number[][] | null
  metrics: Record<string, number>
  trade_list: BacktestTradeRow[]
  is_running: boolean
  last_error: string | null
  tool_extra: [string, string][]
}

File: src/commands/dto/backtester.rs
//! Backtester tab DTOs.

use serde::Serialize;
use std::collections::HashMap;

#[derive(Serialize, Clone)]
pub struct ToolParamDto {
    pub name: String,
    pub value: f64,
    pub min: f64,
    pub max: f64,
    pub step: f64,
}

#[derive(Serialize, Clone)]
pub struct ToolDto {
    pub idx: usize,
    pub name: String,
    pub tooltip: String,
    pub params: Vec<ToolParamDto>,
}

#[derive(Serialize, Clone)]
pub struct StrategyEntryDto {
    pub id: String,
    pub name: String,
    pub domain: String,
}

#[derive(Serialize, Clone)]
pub struct DataSourceDto {
    pub id: String,
    pub name: String,
    pub domain: String,
}

#[derive(Serialize, Clone)]
pub struct BacktestTradeRowDto {
    pub strategy_id: String,
    pub side: String,
    pub entry_price: f64,
    pub exit_price: f64,
    pub size: f64,
    pub pnl: f64,
    pub timestamp: i64,
}

#[derive(Serialize, Clone)]
pub struct BacktesterResultsDto {
    pub equity_curve: Vec<f64>,
    pub drawdown_curve: Vec<f64>,
    pub pnl_buckets: Vec<(f64, u32)>,
    pub mc_paths: Option<Vec<Vec<f64>>>,
    pub metrics: HashMap<String, f64>,
    pub trade_list: Vec<BacktestTradeRowDto>,
    pub is_running: bool,
    pub last_error: Option<String>,
    pub tool_extra: Vec<(String, String)>,
}

#[derive(Serialize, Clone)]
pub struct BacktesterStateDto {
    pub strategies: Vec<StrategyEntryDto>,
    pub data_sources: Vec<DataSourceDto>,
    pub tools: Vec<ToolDto>,
    pub is_running: bool,
}
